### PR TITLE
3-parameter Pinned H2O

### PR DESCRIPTION
**Pull request by siuwuncheung**
This PR sets up the parametric case of Pinned H2O, with 3 parameters (2 bondlengths and 1 bondangle).
Future PR will do parametric ROM. 

----------------------------------------------------------------

**3 samples of 1331 coords.in files from generate_coords_simple.py**

coords_1.00_1.00_0.0.in (Bondlength = {1.83, 1.83}, Bondangle = 104.5)
O1  1   0.00    0.00    0.00  0
H1  2   1.12    1.45    0.00  1
H2  2   1.12    -1.45    0.00  1

coords_1.00_1.00_5.0.in (Bondlength = {1.83, 1.83}, Bondangle = 109.5)
O1  1   0.00    0.00    0.00  0
H1  2   1.06    1.49    0.00  1
H2  2   1.06    -1.49    0.00  1

coords_1.05_0.95_0.0.in (Bondlength = {1.9215, 1.7385}, Bondangle = 104.5)
O1  1   0.00    0.00    0.00  0
H1  2   1.18    1.52    0.00  1
H2  2   1.06    -1.37    0.00  1

----------------------------------------------------------------
**Experiments on rotating a general molecule to the 3-parameter space. Outputs of rotation_test.py**

Original system
H1 = [-0.45  1.42 -1.07]
H2 = [-0.45 -1.48 -0.97]
Bondlength of O1-H1 = 1.8340665200586372
Bondlength of O1-H2 = 1.8258696558078837
Angle between O1-H1 and O1-H2 = 104.90216602748386
Aligned system
H1 = [ 1.19062568e+00  1.39506648e+00 -2.18622350e-16]
H2 = [ 1.03729618e+00 -1.50260329e+00 -2.15581121e-16]
Bondlength of O1-H1 = 1.8340665200586375
Bondlength of O1-H2 = 1.825869655807884
Angle between O1-H1 and O1-H2 = 104.90216602748387

### CODE REVIEW

File: examples/PinnedH2O/generate_coord.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import numpy as np
+import os
+
+# coords.in
+O1 = np.array([0.00, 0.00, 0.00])
+ref_H1 = np.array([-0.45, 1.42, -1.07])
+ref_H2 = np.array([-0.45, -1.48, -0.97])
```
Comment (jeanlucf22) [inline examples/PinnedH2O/generate_coord.py:7]: I would use a symmetric geometry as reference: the equilibrium geometry.

File: examples/PinnedH2O/generate_coord.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    Q_ref_H2 = np.dot(Q, ref_H2)
+    for f_bondlength1 in bondlength1_factor:
+        for f_bondlength2 in bondlength2_factor:
+            H1 = f_bondlength1 * ref_H1
+            H2 = f_bondlength2 * Q_ref_H2
+            filename = f"{output_dir}/coords_{f_bondlength1:.2f}_{f_bondlength2:.2f}_{d_bondangle}.in"
+            with open(filename, "w") as file:
+                file.write(f"O1  1   {O1[0]:.2f}    {O1[1]:.2f}    {O1[2]:.2f}  0\n")
+                file.write(f"H1  2   {H1[0]:.2f}    {H1[1]:.2f}    {H1[2]:.2f}  1\n")
+                file.write(f"H2  2   {H2[0]:.2f}    {H2[1]:.2f}    {H2[2]:.2f}  1\n")
```
Comment (jeanlucf22) [inline examples/PinnedH2O/generate_coord.py:63]: I assumed you verified the resulting coordinates are in the same plane. Is it the z=0 plane?

File: examples/PinnedH2O/generate_coord.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+ref_bondlength1 = calculate_bondlength(ref_H1, O1)
+ref_bondlength2 = calculate_bondlength(ref_H2, O1)
+ref_bondangle = calculate_bondangle(ref_H1, O1, ref_H2)
+
+normal_vector = np.cross(ref_H1, ref_H2)
+normal_unit_vector = normal_vector / np.linalg.norm(normal_vector)
+
+for d_bondangle in bondangle_increment:
+    Q = rotation_matrix(normal_unit_vector, d_bondangle)
```
Comment (jeanlucf22) [inline examples/PinnedH2O/generate_coord.py:53]: The rotation matrix should not depend on the bond angle, but should be determined by the angle between the normal vector and the z=0 plane (if you want to rotate molecule into the z=0 plane)

File: examples/PinnedH2O/generate_coord.py (hunk 1 above)
Comment (siuwuncheung) [reply examples/PinnedH2O/generate_coord.py:7]: I created a new file named `examples/PinnedH2O/generate_coord_simple.py` and it has the symmetry geometry as reference.

File: examples/PinnedH2O/generate_coord.py (hunk 3 above)
Comment (siuwuncheung) [reply examples/PinnedH2O/generate_coord.py:53]: I created a new file named `examples/PinnedH2O/generate_coord_simple.py` and it has the molecule always in the z = 0 plane.

File: examples/PinnedH2O/generate_coord.py (hunk 2 above)
Comment (siuwuncheung) [reply examples/PinnedH2O/generate_coord.py:63]: In the original file, it is always in the same plane, spanned by `ref_H1` and `ref_H2`. With `examples/PinnedH2O/generate_coord_simple.py`, which is much simpler, it is in the z = 0 plane.

**Review by jeanlucf22 — APPROVED** [review]

**Pull request merged**